### ISSUE TITLE
Inconsistent ranges of `CartesianIndex`

### ISSUE THREAD

**Issue by Seelengrab** [JuliaLang/julia#43336]
I'd have expected these two to be equivalent, but the second one errors:

```julia
julia> const ci = CartesianIndex
CartesianIndex

julia> ci(1,1):ci(3,1) |> collect
3×1 Matrix{CartesianIndex{2}}:
 CartesianIndex(1, 1)
 CartesianIndex(2, 1)
 CartesianIndex(3, 1)

julia> ci(1,1):ci(1,0):ci(3,1) |> collect
ERROR: ArgumentError: step cannot be zero
Stacktrace:
 [1] steprange_last(start::Int64, step::Int64, stop::Int64)
   @ Base ./range.jl:325
 [2] StepRange
   @ ./range.jl:312 [inlined]
 [3] StepRange
   @ ./range.jl:367 [inlined]
 [4] _colon
   @ ./range.jl:24 [inlined]
 [5] Colon
   @ ./range.jl:22 [inlined]
 [6] #7
   @ ./multidimensional.jl:318 [inlined]
 [7] map (repeats 2 times)
   @ ./tuple.jl:266 [inlined]
 [8] (::Colon)(I::CartesianIndex{2}, S::CartesianIndex{2}, J::CartesianIndex{2})
   @ Base.IteratorsMD ./multidimensional.jl:317
 [9] top-level scope
   @ REPL[3]:1
```

Unintuitively, the "correct" way to write this would be `ci(1,1):ci(1,x):ci(3,1)` for any value `x` other than `0`:

```julia
julia> ci(1,1):ci(1,1):ci(3,1) |> collect
3×1 Matrix{CartesianIndex{2}}:
 CartesianIndex(1, 1)
 CartesianIndex(2, 1)
 CartesianIndex(3, 1)

julia> ci(1,1):ci(1,10):ci(3,1) |> collect
3×1 Matrix{CartesianIndex{2}}:
 CartesianIndex(1, 1)
 CartesianIndex(2, 1)
 CartesianIndex(3, 1)
```

I think this is a consequence of how the cartesian product is formed right now - maybe dimensions where the step is zero can be just skipped during product expansion, provided both `start` and `stop` of the to-be-constructed range have the same coordinate in that point? Since no dimensions are lost, it should still be type stable.

**Comment by stev47** [comment]
The default step-size for ranges is one, even if start and end agree. In your example the step-size in the second dimension should be one, not zero.
The error boils down to step-ranges not allowing a zero step-size (try `1:0:1`), which is a useful convention (think e.g. of length calculation). See also [this discussion](https://discourse.julialang.org/t/range-step-cannot-be-zero/66948) for further details.

**Comment by johnnychen94** [comment]
See also the discussions in https://github.com/JuliaLang/julia/issues/41959 and https://github.com/JuliaLang/julia/pull/41960

There are two valid interpretations of `CartesianIndex{N}` here:

- a point (x¹, x², ..., xᴺ)` in N-dimensional space: pⁱ = p₀ⁱ + xⁱ * Δⁱ where pⁱ are real numbers.
- a point (x,) in 1-dimensional space: p = p₀ + x * Δ where x is real number.

`ci(1,1):ci(3,1)` goes the first interpretation thus the output is a 2-D matrix instead of a 1-D vector.

**Comment by Seelengrab** [comment]
@stev47 Can you please not just close an issue where I already mention that this error is due to the assumption that the cartesian product of each dimension "just works" for the expansion over a nonzero step? The discourse discussion you're linking is also not relevant here - `CartesianIndex(1,0)` is decidedly _not_ `zero(CartesianIndex{N})` for any `N`.

@johnnychen94 Yes, I've read & participated in both of those discussions. I don't think they apply here. My complaint is not with `a:b`, but with `a:b:c` behaving wrongly in the case that `b` has a step of `0` in one direction, but otherwise does create elements. Specifically I think it would be nice if `CartesianIndex(1,1):CartesianIndex(1,0):CartesianIndex(10,1) == `CartesianIndex(1,1):CartesianIndex(10,1)` or `CartesianIndex(1,1):CartesianIndex(0,1):CartesianIndex(1,10) == `CartesianIndex(1,1):CartesianIndex(1,10)` would both hold.

In short, I think this should only error if the step is `0` in all dimensions (i.e. the step is `zero(CartesianIndex{N})`). In other words, dimensions in the step that are `0` must be equal in both `start` & `stop`, otherwise build the cartesian product over the nonzero dimensions with extent in start & stop. This does not loose dimension information.

**Comment by Seelengrab** [comment]
Maybe let me expand on the motivation here - creating a cartesian product is not my main goal. I'm mainly interested in getting steps like `CartesianIndex(1,0,0,0)` or `CartesianIndex(0,0,5,0)` to work, since they're very useful when wanting to create a line along a given axis. I'm fully aware that this will not and cannot create a line that's off-axis (e.g. with step like `CartesianIndex(1,0,3,0)`). That is not the goal here.

**Comment by johnnychen94** [comment]
@Seelengrab I see your point now. We should revive the discussion https://github.com/JuliaLang/julia/issues/15218 first; once `1:x:1` is supported then it's immediately followed that `CartesianIndex(1, 1):CartesianIndex(1,0):CartesianIndex(4, 1)` is also supported. For the moment let me reopen this issue.

**Comment by Seelengrab** [comment]
I'm not sure we have to solve the general case mentioned in that issue for this to be possible - simply ignoring the dimensions with `0` as their step during the product would already be consistent with the cartesian-ness of the `CartesianIndex` range, no? After all, the dimension isn't lost since all points are then required to have the same coordinate in that dimension. It just errors right now because the code assumes that this is the same as a cartesian product over the individual stepranges, which imo is a different case and would only matter when the step is `zero(CartesianIndex{N})`.

**Comment by johnnychen94** [comment]
> simply ignoring the dimensions with 0 as their step during the product would already be consistent with the cartesian-ness of the CartesianIndex range, no?

Yes, that's a shortcut but I think it's essentially no different to the general 1-d range case. I understand your point that `CartesianIndex` is more useful a case that we may want to support; maybe you can put up a PR and get more serious feedback from others? It should be quite easy, just needs some effort on docstrings and tests. 

```julia
# something like this :)
function Base.:(:)(I::CartesianIndex{N}, S::CartesianIndex{N}, J::CartesianIndex{N}) where N
    R = map(Tuple(I), Tuple(S), Tuple(J)) do i, s, j
        if iszero(s) && i == j
            i:one(i):i
        else
            i:s:j
        end
    end
    CartesianIndices(R)
end
```

**Comment by stev47** [comment]
@Seelengrab sorry for being hasty but no harm was done to you.
@johnnychen94 thank you for linking #15218, that discussion is interesting and I can see some argument for allowing zero-steps in general.
I still think that both issues are linked, since one would expect `1:0:1` and `CartesianIndex(1):CartesianIndex(0):CartesianIndex(1)` to behave similarly.

**Comment by Seelengrab** [comment]
> sorry for being hasty but no harm was done to you.

No worries! Just felt very dismissive, as I've explicitly mentioned that I'm concerned with nonzero steps. Maybe I could have phrased that more clearly in the initial post.

> I still think that both issues are linked, since one would expect 1:0:1 and CartesianIndex(1):CartesianIndex(0):CartesianIndex(1) to behave similarly.

That is orthogonal to this issue though, as `CartesianIndex(0)` truly is `zero(CartesianIndex{1})`. I'm explicitly talking about cases where at least one dimension _isn't_ zero, making sure the step is never `zero` of the relevant type :)

**Comment by stev47** [comment]
I see, then it is indeed more viable.
It would however mean, that they no longer respect the tensor product structure, i.e. for tuples `a`, `b` and `s`:
```julia
CartesianIndex(a):CartesianIndex(s):CartesianIndex(b) ==
  CartesianIndex.(Iterators.product(map((a, s, b) -> a:s:b, a, s, b)...))
```
Not sure how valuable that consistency is.

**Comment by mbauman** [comment]
This is quite unfortunate — for more than just the cases in the first post — but I can see how we got here.

The CartesianIndex `start:stop` **isn't** a one-dimensional vector with a step size equal to one (as the generic form of `:` means)... but yet it kinda makes some sort of sense to me and I think it's quite defensible. If we think about ranges as sequential additions, it's stepping (adding) a `CartesianIndex((1, 0, 0, 0, ...))` _with a higher-dimensional "carry"_ for the given shape.  That is, the second element is not the result of adding `1`, nor is it adding `one`, nor is it adding `oneunit`.  It's doing something special and multidimensional.  And with a special docstring (as we have) I think it's great.

But once we have that special `start:stop` method, what does `start:step:stop` mean?  Somehow passing a particular `step` _really_ makes me think that it should be sequential additions.  The second element should be `start + step`.  The third should be `start + step + step`.  But instead we propagated the special multidimensional case — thinking not in terms of whole-`CartesianIndex` addition, but rather stepping _within_ their independent dimensions.

This has gotten worse because we've simultaneously approached this with the opposite interpretation within `StepRangeLen` — that _actually does_ add the `step`.  So now we have:

```Julia
julia> collect(CartesianIndex(1,1):CartesianIndex(3,3))
3×3 Matrix{CartesianIndex{2}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)  CartesianIndex(1, 3)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)  CartesianIndex(2, 3)
 CartesianIndex(3, 1)  CartesianIndex(3, 2)  CartesianIndex(3, 3)

julia> collect(CartesianIndex(1,1):CartesianIndex(1,1):CartesianIndex(3,3))
3×3 Matrix{CartesianIndex{2}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)  CartesianIndex(1, 3)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)  CartesianIndex(2, 3)
 CartesianIndex(3, 1)  CartesianIndex(3, 2)  CartesianIndex(3, 3)

julia> collect(range(start=CartesianIndex(1,1), step=CartesianIndex(1,1), stop=CartesianIndex(3,3)))
3×3 Matrix{CartesianIndex{2}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)  CartesianIndex(1, 3)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)  CartesianIndex(2, 3)
 CartesianIndex(3, 1)  CartesianIndex(3, 2)  CartesianIndex(3, 3)

julia> collect(StepRangeLen(CartesianIndex(1,1), CartesianIndex(1,1), 3))
3-element Vector{CartesianIndex{2}}:
 CartesianIndex(1, 1)
 CartesianIndex(2, 2)
 CartesianIndex(3, 3)
```

The collision of these two views comes to a fore in `range`.  And so the `length` kwarg was made to be an error with #50457.

I wish we could walk back the `:` and `range` with explicit step here, but that's been happening since 1.6 and is documented.

**Comment by Seelengrab** [comment]
Sigh.. So there are more inconsistencies now than there were when I opened this issue? That makes me sad :(


Conceptually, I'm quite convinced now that `range` and `StepRangeLen` should behave the same, but alas, it seems this ship too has sailed now..

**Comment by adienes** [comment]
I ordinarily dislike "magic" but I actually think the `a:b` rectangular behavior here is kind of cool. I agree with your summary that it looks like the ideal solution would be to ~have never added~ remove the ability to specify `step` since that must be ambiguous

**Comment by mbauman** [comment]
If the `CartesianIndex(...):CartesianIndex(...)` returned a `vec(CartesianIndices(...))` **and** we didn't define `oneunit`, then I think it's _almost_ completely consistent with the generic definition and wouldn't need a special docstring at all.  The only strange bit is that it implicitly assumes that the `start` and `stop` are themselves the extents you care about.  E.g., you could imagine traversing column major from `CartesianIndex(4, 1)` to `CartesianIndex(4, 2)` within the context of a `6x2` array (that is, linear indices `4:10`).  But `CartesianIndex(4, 1):CartesianIndex(4, 2)` just hits that row.

Perhaps most tellingly, though, the `StepRangeLen` interpretation here is completely consistent with the generic definition.
